flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -835,27 +835,65 @@ class ExprLowering {
       return array;
     } else {
-      // Convert Ev::ConstantSubs to SequenceType::Shape
       fir::SequenceType::Shape shape(con.shape().begin(), con.shape().end());
       auto eleTy = converter.genType(TC, KIND);
       auto arrayTy = fir::SequenceType::get(shape, eleTy);
       auto idxTy = builder.getIndexType();
       mlir::Value array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
       Fortran::evaluate::ConstantSubscripts subscripts = con.lbounds();
+      bool foundRange = false;
+      mlir::Value rangeValue;
+      llvm::SmallVector<mlir::Value, 8> rangeStartIdx;
+
       do {
         auto constant =
             fir::getBase(genScalarLit<TC, KIND>(con.At(subscripts), con));
         llvm::SmallVector<mlir::Value, 8> idx;
         for (const auto &pair : llvm::zip(subscripts, con.lbounds())) {
-          const auto &dim = std::get<0>(pair);
-          const auto &lb = std::get<1>(pair);
+          const auto &[dim, lb] = pair;
+
+          // Add normalized upper bound index to idx.
           idx.push_back(
               builder.createIntegerConstant(getLoc(), idxTy, dim - lb));
         }
+
         auto insVal = builder.createConvert(getLoc(), eleTy, constant);
-        array = builder.create<fir::InsertValueOp>(getLoc(), arrayTy, array,
-                                                   insVal, idx);
+        auto nextSubscripts(subscripts);
+
+        // Check to see if the next value is the same as the current value
+        bool nextIsSame = con.IncrementSubscripts(nextSubscripts) &&
+                          con.At(subscripts) == con.At(nextSubscripts);
+        bool newRange = (nextIsSame != foundRange) && !foundRange;
+        bool endOfRange = (nextIsSame != foundRange) && foundRange;
+        bool continueRange = nextIsSame && foundRange;
+
+        if (newRange) {
+          // Mark the start of the range
+          rangeStartIdx = idx;
+          rangeValue = insVal;
+          foundRange = true;
+        } else if (endOfRange) {
+          // Zip together the upper and lower bounds of the range for each
+          // index in the form [lb0, up0, lb1, up1, ... , lbn, upn] to pass
+          // to the InserOnEangeOp.
+          llvm::SmallVector<mlir::Value, 8> zippedRange;
+          for (size_t i = 0; i < idx.size(); ++i) {
+            zippedRange.push_back(rangeStartIdx[i]);
+            zippedRange.push_back(idx[i]);
+          }
+          array = builder.create<fir::InsertOnRangeOp>(getLoc(), arrayTy, array,
+                                                       rangeValue, zippedRange);
+          foundRange = false;
+        } else if (continueRange) {
+          // Loop until the end of the range is found.
+          continue;
+        } else /* no range */ {
+          // If a range has not been found then insert the current value.
+          array = builder.create<fir::InsertValueOp>(getLoc(), arrayTy, array,
+                                                     insVal, idx);
+        }
```

Comment:
Does this find run lengths inter-dimensionally? It looks like it's just scoring a sequence if it appears in the linear address form.  Say we have
```fortran
block data
  real :: arr(5,5)
  common /block/ arr
  data x(1,1), x(2,1), x(3,1) / 1, 1, 0 /
  data x(1,2), x(2,2), x(4,2) / 1, 1, 2.4 /
  data x(1,3), x(2,3), x(4,3) / 1, 1, 2.4 /
  data x(4,4) / 2.4 /
end
```
The logical sub matrix `[1..2;1..3]` has value 1.0 and the logical subvector `[4;2..4]` has a value 2.4 (and there are subspaces of all 0.0), but clearly these values are not linearly contiguous.

Reply:
Indeed. Do you think that it would be reasonable to try to lower something like this directly into using `InsertOnRangeOp`? It seems more like something that would be more easily caught by a transformation over `fir::InsertValueOp`s.

Reply:
I'm not sure of the utility of the general _n_ dimensional form ... I guess I'm not sure what the semantics of this range op are now.

Reply:
That is valid, I suppose I don't actually know if there are any optimizations that can benefit/utilize `InsertOnRange` or if it is entirely for shrinking the mlir code.

Reply:
Originally, the idea was to compactly encode something like:
```fortran
   REAL :: VastArrayOfValues(1000,1000,1000) = 1.0
```
so that 1,000,000,000 insert value instructions were not created.